File: PriorityQueue/1514_Path_with_Maximum_Probability/1514_Path_with_Maximum_Probability.cpp
class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start_node, int end_node) {
        
        // Adjusted adjacency list to hold vectors of pairs
        unordered_map<int, vector<pair<int, double>>> adj;

        // Populate the adjacency list
        for(size_t i = 0; i < edges.size(); i++) {
            int src = edges[i][0];
            int dst = edges[i][1];
            adj[src].emplace_back(dst, succProb[i]);
            adj[dst].emplace_back(src, succProb[i]); // Assuming it's an undirected graph
        }

        // Priority queue to process nodes by highest probability
        priority_queue<pair<double, int>> pq;
        pq.push({1.0, start_node}); // start with probability 1.0 at the start node

        // Vector to store the maximum probability to reach each node
        vector<double> maxProb(n, 0.0);
        maxProb[start_node] = 1.0;

        while(!pq.empty()) {
            auto [prob, curr] = pq.top();
            pq.pop();

            if (curr == end_node) {
                return prob;  // Return the probability when reaching the end node
            }

            for (auto& [nei, edgeProb] : adj[curr]) {
                double newProb = prob * edgeProb;

                if (newProb > maxProb[nei]) {  // Only consider this path if it provides a higher probability
                    maxProb[nei] = newProb;
                    pq.push({newProb, nei});
                }
            }
        }

        return 0.0; // If there's no path found
    }

};

File: PriorityQueue/1514_Path_with_Maximum_Probability/1514_Path_with_Maximum_Probability.java
class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start, int end) {
        // Adjacency list: each node maps to a list of pairs (neighbor, edge probability)
        Map<Integer, List<Pair<Integer, Double>>> adj = new HashMap<>();

        // Build the graph
        for (int i = 0; i < edges.length; i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            double prob = succProb[i];

            adj.computeIfAbsent(u, k -> new ArrayList<>()).add(new Pair<>(v, prob));
            adj.computeIfAbsent(v, k -> new ArrayList<>()).add(new Pair<>(u, prob));
        }

        // Priority queue for Dijkstra's algorithm (max-heap based on probability)
        PriorityQueue<Pair<Double, Integer>> pq = new PriorityQueue<>((a, b) -> Double.compare(b.getKey(), a.getKey()));
        pq.offer(new Pair<>(1.0, start));

        // Array to store the max probability to reach each node
        double[] maxProb = new double[n];
        maxProb[start] = 1.0;

        // Dijkstra's algorithm
        while (!pq.isEmpty()) {
            Pair<Double, Integer> current = pq.poll();
            double prob = current.getKey();
            int node = current.getValue();

            if (node == end) return prob; // Early exit if end node is reached

            // Iterate over neighbors
            for (Pair<Integer, Double> neighbor : adj.getOrDefault(node, new ArrayList<>())) {
                int nextNode = neighbor.getKey();
                double edgeProb = neighbor.getValue();
                double newProb = prob * edgeProb;

                if (newProb > maxProb[nextNode]) {
                    maxProb[nextNode] = newProb;
                    pq.offer(new Pair<>(newProb, nextNode));
                }
            }
        }

        return 0.0; // If no path exists
    }

    // Pair class to hold node and probability
    class Pair<K, V> {
        private K key;
        private V value;

        public Pair(K key, V value) {
            this.key = key;
            this.value = value;
        }

        public K getKey() {
            return key;
        }

        public V getValue() {
            return value;
        }
    }
}